src/cpp/session/SessionMain.cpp

```diff
@@ -3243,11 +3243,27 @@ int main (int argc, char * const argv[])
       error = workingDir.makeCurrentPath();
       if (error)
          return sessionExitFailure(error, ERROR_LOCATION);
-         
-      // start http connection listener
-      error = startHttpConnectionListener();
-      if (error)
-         return sessionExitFailure(error, ERROR_LOCATION);
+      
+      // Wait for port to be ready, if needed. This triggers when the
+      // rsession restarts but takes the port a few ms to become available
+      int retries = 10;
+      while (retries > 0)
+      {
+          // start http connection listener
+          error = startHttpConnectionListener();
+          if (error)
```

Comment:
I agree we should restrict the retry to Address already in use. You can see examples of testing for specific error types here: https://github.com/rstudio/rstudio/blob/master/src/cpp/core/include/core/http/SocketUtils.hpp#L69-L74